{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"data_generation.ipynb","provenance":[],"collapsed_sections":[],"mount_file_id":"1mEvlnuj-ThOS50fc2ID_o_gz64wgka-j","authorship_tag":"ABX9TyPNP5V57HNhoqCeQPjGs/Ex"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"rjOsulZJLuTf"},"outputs":[],"source":["# pysam installation\n","!pip install pysam"]},{"cell_type":"code","source":["import pickle\n","import os\n","import pysam\n","import numpy as np"],"metadata":{"id":"rIWiQ-HUBmGr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**Extracting 289 bins of size 10 Mb from whole genome**"],"metadata":{"id":"5wv5FpbbIun9"}},{"cell_type":"code","source":["parentPath = os.getcwd()\n","\n","contig_windows = {}\n","with open('meta_info/contig_windows.pkl', 'rb') as handle:\n","  contig_windows = pickle.load(handle)\n","\n","our_contigs = []\n","for i in range(1, 23):\n","  our_contigs.append('chr'+str(i))"],"metadata":{"id":"E3TKEgS3B4Y6"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Helper functions for sample generation\n","def init_len_dict():\n","  len_dict = {}\n","  for i in range(1, 401):\n","    len_dict[i] = 0\n","  return len_dict\n","\n","def dict_to_list(dictt):\n","  lst = []\n","  for i in range(1, 401):\n","    lst.append(dictt[i])\n","  return lst\n","\n","# traversing bam file and filling up the windows based on length frequency\n","def sample_generation(bamPath): \n","  sample = []\n","  len_dict = init_len_dict()\n","  pybam = pysam.AlignmentFile(bamPath, \"rb\")\n","  for contig in our_contigs: # genome\n","    range_list = contig_windows[contig]\n","    for lst in range_list: # contig\n","      start, end = lst[0], lst[1]\n","      for read in pybam.fetch(contig, start, end): # window\n","        length = read.template_length\n","        if length<=0 or length>400:\n","          continue\n","        else:\n","          len_dict[length] += 1\n","      sample.append(dict_to_list(len_dict))\n","      len_dict = init_len_dict()\n","  \n","  sample_arr = np.array(sample).astype(np.int16)\n","  return sample_arr"],"metadata":{"id":"Aafxi2JMC1N4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["os.chdir('bam_files')\n","\n","sample_cnt = 0\n","for folder in os.listdir():\n","  sample_cnt += 1\n","data = np.zeros((sample_cnt, 289, 400))\n","\n","i = 0\n","sample_names = []\n","for folder in os.listdir():\n","  data[i] = sample_generation(folder + '/sample.bam')\n","  sample_names.append(folder)\n","  i += 1\n","\n","# Removing 7 black listed bins with no fragments from the 289 bins\n","no_read_windows = [13, 159, 209, 221, 232, 233, 284]\n","read_windows = []\n","for i in range(289):\n","  if i not in no_read_windows:\n","    read_windows.append(i)\n","data = data[:, read_windows, :]"],"metadata":{"id":"B6wg_4aVDTFL"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**Feature extraction and saving as intermediate files to be used by the models**"],"metadata":{"id":"rlBzIN5bJY_V"}},{"cell_type":"code","source":["os.chdir(parentPath)\n","sig_dict = {}\n","with open('meta_info/sig_bins.pickle', 'rb') as handle:\n","  sig_dict = pickle.load(handle)\n","sig_lens = sig_dict['sig_lens']\n","\n","gene_bin_dict = {}\n","with open('meta_info/gene_bin.pkl', 'rb') as handle:\n","  gene_bin_dict = pickle.load(handle)\n","tumor_bins = gene_bin_dict['cancer']\n","blood_bins = gene_bin_dict['blood']"],"metadata":{"id":"sylyEn8JG4qk"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["high_data = np.sum(data[:, :, 50:], axis=1)\n","\n","low_data = data[:, :, 50:]\n","low_data = low_data[:, :, sig_lens]\n","low_tumor = np.expand_dims( np.sum(low_data[:, tumor_bins, :], axis=1), axis=1)\n","low_blood = np.expand_dims( np.sum(low_data[:, blood_bins, :], axis=1), axis=1)\n","low_data = None\n","low_data = np.concatenate((low_tumor, low_blood), axis=1)"],"metadata":{"id":"Iz3razr5ZPHW"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["low_dict = {}\n","low_dict['samples'] = low_data\n","low_dict['meta_info'] = sample_names\n","os.chdir(parentPath)\n","with open('Intermediate_Files/low_data.pkl', 'wb') as f:\n","  pickle.dump(low_dict, f)"],"metadata":{"id":"WqaEPx5yeNyH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["high_dict = {}\n","high_dict['samples'] = high_data\n","high_dict['meta_info'] = sample_names\n","os.chdir(parentPath)\n","with open('Intermediate_Files/high_data.pkl', 'wb') as f:\n","  pickle.dump(high_dict, f)"],"metadata":{"id":"x3tAz6XpfYNk"},"execution_count":null,"outputs":[]}]}